### ISSUE TITLE
Relative URIs don’t work in annotated playlists

### ISSUE THREAD

**Issue by Moonbase59** [savonet/liquidsoap#3741]
**Describe the bug**
I have all my playlists using relative URIs, example `Test.m3u`:

```
../../Other/Jingles/test-5-15-5-15-5s.mp3
```
and they play just fine.

When I try to _annotate_ entries, example: `Test-processed.m3u8`:
```
#EXTM3U
annotate:liq_cue_in="5.00",liq_cross_duration="4.69",duration="44.99",liq_amplify="0.50dB":../../Other/Jingles/test-5-15-5-15-5s.mp3
```

I get an `Nonexistent file or ill-formed URI "../../Other/Jingles/test-5-15-5-15-5s.mp3"!` error.

**To Reproduce**
```
uri = "~/Musik/Playlists/Radio/Test-processed.m3u8"
radio = playlist(uri)

def live_aware_crossfade(old, new) =
    log(label="crossfade", "Using custom crossfade")
    cross.smart(old, new, fade_in=0.1, fade_out=2.5, margin=8.)
end

radio = cross(duration=3.00, width=2.00, live_aware_crossfade, radio)

radio = mksafe(radio)

output.ao(radio)
```

**Expected behavior**
Relative URIs working with annotated playlist files.

**Version details**

- OS: Linux Mint 21.3
- Version: Liquidsoap 2.2.4-1

**Install method**
.deb

**Comment by vitoyucepi** [comment]
Hi @Moonbase59,
Could you write an absolute path for these files?

> `~/Musik/Playlists/Radio/Test-processed.m3u8`
> `../../Other/Jingles/test-5-15-5-15-5s.mp3`

Also, what's the current working directory for the liquidsoap script?

**Comment by vitoyucepi** [comment]
I asked you to manually resolve files for reproduction purposes only.

I can confirm that liquidsoap resolves files in annotated playlists relative to CWD, not the playlist file itself.

**Comment by Moonbase59** [comment]
No offense. Just tried to explain that these are on a NAS and mounted from many different machines, on different  mount points, so relative paths were the choice. Of course I did convert a few just to test. :-)

Think it would be possible to fix this, i.e. have them resolve to the location of the playlist, much like "normal" playlists work? Thinking of embedded Liquidsoap, like in AzuraCast, it might not always be possible to change the CWD to the location of the playlist…

**Comment by toots** [comment]
From a system point of view, I don't think that it is reasonable to expect much from relative URIs. 

Your assumption that it should be referering to the playlist file is not the usual one. in fact. In most situations, relatives URIs are resolved relative the _current working directory_ (CWD) which is the directory from where the user is typing its command.

In general it cannot be expected that all URIs come from a single playlist. URIs can for instance be pushed individually to a `request.queue`. Therefore, given a URI of the form: `annotate:<metadata>:../../bla.mp3` there is no generic notion of playlist attached to it.

In some specific simple case, such as a simple text playlist with relative URIs, it might have been possible to pre-process each URI to resolve the relative URI based on the base playlist. However, as soon at the URIs start being more complex such as one with `annotate`, then getting an actual file path requires applying the various required protocol resolutions to that URIs.

When that happens, the URI is being processed through a generic protocol resolution making it impossible to resolve relative URIs in a consistent way except from using the CWD.

Lastly, the notion of current working directory is essentially a terminal notion, relevant from the perspective of a user being logged into a terminal.

When setting up an automated system, it is not wise to rely on this notion.

Typically, a daemonization utility such as `systemd` will most likely change the path from where the script is executed. In fact, the first step of daemonizing removes all traces of shell/terminal, removing TTYs, etc etc. I would expect the current working directory to be gone by then or to be a trivial location because of that and for security reasons.

In conclusion, if you expect your system to be robust, I would absolutely recommend _against_ using relative URIs anywhere.

It seems to me that your initial remark is because of a lack of referential consistency: a URI with relative path is resolved relative to a playlist when the URI is just the path but not when it is prefixed. As explained above, this is a discrepancy in the resolving conditions. If anything should be done about on liquidsoap side, I would suggest in this case to _remove_ the simple relative URI pre-processing so as to not confuse the use about what will happen when their simple URIs/playlist start being used in a more comple context with protocols and prefixes.
